src/kem/kem.h

```diff
@@ -177,6 +193,21 @@ typedef struct OQS_KEM {
 	 */
 	OQS_STATUS (*keypair)(uint8_t *public_key, uint8_t *secret_key);
 
+	/**
+	 * Derandomized encapsulation algorithm.
+	 *
+	 * Caller is responsible for allocating sufficient memory for `ciphertext` and
+	 * `shared_secret`, based on the `length_*` members in this object or the per-scheme
+	 * compile-time macros `OQS_KEM_*_length_*`.
+	 *
+	 * @param[out] ciphertext The ciphertext (encapsulation) represented as a byte string.
+	 * @param[out] shared_secret The shared secret represented as a byte string.
+	 * @param[in] public_key The public key represented as a byte string.
+     * @param[in] coins The input randomness represented as a byte string.
+	 * @return OQS_SUCCESS or OQS_ERROR
+	 */
+	OQS_STATUS (*encaps_derand)(uint8_t *ciphertext, uint8_t *shared_secret, const uint8_t *public_key, const uint8_t *coins);
```

Comment:
OK -- now I understand some more: The underlying implementation asks for the seed/coins to be passed in each call. This however begs the question: How many "coins" are needed to correctly operate this API? Shouldn't this be documented & checked here? Equivalent question: How many coins are needed for the keygen API call? Beyond the ability to improve API robustness (check this at invocation), if this were known(?) wouldn't it be possible to pass a sufficient number of "coins" to a "setup_random_coins" call that is invoked once and then used "under the covers" for each subsequent call to (API-wise unchanged) keygen and encaps calls? That way the many "OQS_ERROR" return responses for the KEM algs not supporting that could be saved: They simply wouldn't use "prepared coins" when asked to operate keygen&encaps.

Reply:
Sorry it's taken me so long to follow up on this, Michael.

For the first question, `length_keypair_coins` and `length_encaps_coins` are added to the `OQS_KEM` struct; the caller can use them similarly to `length_shared_secret`, `length_ciphertext`, etc.

Let me confirm I'm understanding correctly: `setup_random_coins` would be a one-time call to set up (i.e., seed) a source of randomness which would then be used as RNG for `OQS_randombytes` in subsequent KEM operations? This would be similar in functionality to the `OQS_randombytes_custom_algorithm` call but wouldn't require the caller to define their own function and pass it as a pointer—correct?

The purpose of this PR is to achieve something different. The `keypair_derand` call is intended to implement (well, expose) the [DeriveKeyPair](https://www.rfc-editor.org/rfc/rfc9180.html#name-derivekeypair) KEM function defined in RFC 9180. This is used notably in the recent [Messaging Layer Security](https://www.rfc-editor.org/rfc/rfc9420.html#name-cipher-suites) standard. By design, this is basically a KDF for KEM keys—the caller provides input keying material and gets back a KEM keypair—which needs to be called multiple times rather than having a one-time setup. The intent is to supplement the existing keygen function, not replace it: RFC 9180 requires that the KEM supports [both](https://www.rfc-editor.org/rfc/rfc9180.html#section-4-2.1.1) operations. ML-KEM already supports this operation as part of its API, as well as a similar `encaps_derand` function. (I don't know an application for the latter off the top of my head.) I imagine it would have been nice to have this background before you looked at the PR; sorry for the poor coordination on my part.

Reply:
Thanks for the explanations. I now understand some more -- but clearly not everything: I thought this PR is meant to enable https://github.com/open-quantum-safe/oqs-provider/issues/447? If this were so (?), I'm afraid this API will require substantial changes to `oqsprovider` to facilitate this, no? But then again, maybe you or @Eddy-M-K can show me wrong by doing a simple PR implementing https://github.com/open-quantum-safe/oqs-provider/issues/447 based on this new API (maybe a "tracking-..." branch to trigger CI as we also do with new algs?). Again, will be online only by smart phone until Aug 19, so please do not expect serious further feedback from me until then.